core/src/contracts/stablex_contract/search_batches.rs

```diff
@@ -0,0 +1,202 @@
+// NOTE: Required for automock.
+#![cfg_attr(test, allow(clippy::ptr_arg))]
+
+use anyhow::{anyhow, Result};
+use ethcontract::{prelude::Web3, transport::DynTransport, web3::types::Block, BlockNumber, H256};
+use futures::{
+    compat::Future01CompatExt as _,
+    future::{BoxFuture, FutureExt as _},
+};
+#[cfg(test)]
+use mockall::automock;
+
+fn get_block_batch_id<T>(block: &Block<T>) -> u32 {
+    const BATCH_DURATION: u64 = 300;
+    (block.timestamp.as_u64() / BATCH_DURATION) as _
+}
+
+async fn get_block(
+    web3: &Web3<DynTransport>,
+    block_number: BlockNumber,
+) -> Result<ethcontract::web3::types::Block<H256>> {
+    web3.eth()
+        .block(block_number.into())
+        .compat()
+        .await?
+        .ok_or_else(|| anyhow!("block {:?} is missing", block_number))
+}
+
+#[cfg_attr(test, automock)]
+pub trait BatchIdRetrieving {
+    fn batch_id_from_block<'a>(&'a self, block_number: BlockNumber) -> BoxFuture<'a, Result<u32>>;
+
+    fn current_batch_id_and_block_number<'a>(&'a self) -> BoxFuture<'a, Result<(u32, u64)>>;
+}
+
+impl BatchIdRetrieving for Web3<DynTransport> {
+    fn batch_id_from_block<'a>(&'a self, block_number: BlockNumber) -> BoxFuture<'a, Result<u32>> {
+        async move {
+            let current_block = get_block(&self, block_number).await?;
+            Ok(get_block_batch_id(&current_block))
+        }
+        .boxed()
+    }
+
+    fn current_batch_id_and_block_number<'a>(&'a self) -> BoxFuture<'a, Result<(u32, u64)>> {
+        async move {
+            let current_block = get_block(&self, BlockNumber::Latest).await?;
+            let batch_id = get_block_batch_id(&current_block);
+            let block_number = current_block
+                .number
+                .ok_or_else(|| {
+                    anyhow!("latest block {:?} has no block number", current_block.hash)
+                })?
+                .as_u64();
+            Ok((batch_id, block_number))
+        }
+        .boxed()
+    }
+}
+
+pub async fn search_last_block_for_batch(
+    batch_id_retrieving: &impl BatchIdRetrieving,
+    batch_id: u32,
+) -> Result<u64> {
+    struct Bounds {
+        lower: u64,
+        upper: u64,
+    }
+    impl Bounds {
+        fn diff(&self) -> u64 {
+            self.upper - self.lower
+        }
+        fn mid(&self) -> u64 {
+            (self.upper + self.lower) / 2
+        }
+    }
```

Comment:
Nope! I put it inside the scope at first because it was unrelated to everything else in the code, and I didn't think of moving it out after moving the code to a new file. I also like it more outside of the scope as it is now.